### ISSUE TITLE
symbol-description: allow feature identifier tagging

### ISSUE THREAD

**Issue by Lcfvs** [eslint/eslint#12540]
**What rule do you want to change?**
symbol-description

**Does this change cause the rule to produce more or fewer warnings?**
Fewer, if the empty symbol is a key of an object and it's value is another symbol, for example, used to tag an object by its provider id (the module providing the feature).

It can provide a way, for developers, to easily identify where the features on an object come from, for debugging purpose, without any conflict risks.

**How will the change be implemented? (New option, new default behavior, etc.)?**
new default behavior

**Please provide some example code that this change will affect:**

```js
export default {
  ...featurePrototype,
  [Symbol()]: Symbol(myModuleId)
}
```

**What does the rule currently do for this code?**
Triggers an error.

**What will the rule do after it's changed?**
Ignore in case of.

**Are you willing to submit a pull request to implement this change?**
If needed.

**Comment by mdjermanovic** [comment]
Hi @Lcfvs, thanks for the enhancement proposal!

Wouldn't something like `[Symbol("debug")]: Symbol(myModuleId)` be more explicit and still without conflict risks?

Or just `[Symbol("debug")]: myModuleId`

**Comment by Lcfvs** [comment]
Hi @mdjermanovic I thought about it, as some other ideas, like `@id`,  but I think an empty symbol is better, because the developers can actually store any named symbol, risking to make a conflict.

For examples, a debugging library can export a `debug` symbol, a schema.org library can export a `@id` symboln etc.

The empty seems, by the default rule, a good exception, it isn't?

**Comment by Lcfvs** [comment]
Imho, this special case doesn't need to be explicit, since it's only for a tagging purpose.

**Comment by mdjermanovic** [comment]
Same descriptions shouldn't cause any conflicts in the code?

```js
Symbol("debug") === Symbol("debug"); // false
```

I still think it would be useful to provide a description as it's a feature designed for debugging, even if it happens that some library uses the exact same description for its own symbols.

**Comment by Lcfvs** [comment]
Yeah, but if you have an object like this
```js
{
  [Symbol('debug')]: Symbol(myModuleId),
  ...{// the prototype
    [Symbol('debug')]: myDebugFeature
  }
}
```
It's okay, the symbols are different, but we don't have, at console reading, a way to distinct them.

**Comment by Lcfvs** [comment]
By "conflict", I don't mean in code, just at reading ;)

**Comment by ljharb** [comment]
`Symbol()` and `Symbol()` also don't provide any way to distinguish between them. So far your use case seems like every use should have a distinct description, and none of them should have none.

**Comment by Lcfvs** [comment]
@ljharb As a feature module "meta" symbol, it's exactly the idea, we don't need to distinct them, except the meaning "this symbol isn't a real property, it's just an informative purpose".

```js
export default {
  [Symbol()]: Symbol('a')
  /* , ... features /*}
}
```

```js
export default {
  [Symbol()]: Symbol('b')
  /* , ... features /*}
}
```

```js
import a from './a.js'
import b from './b.js'

console.log({ ...a, ...b }) // Object { Symbol(): Symbol(a), Symbol(): Symbol(b)/* , ... features /*} }
```

**Comment by Lcfvs** [comment]
By these empty symbols (meta), we can easily identify what modules are used to construct the resulting object.

**Comment by ljharb** [comment]
Then make the description something that says that, instead of introducing the need for your devs to remember an arbitrary convention.

**Comment by Lcfvs** [comment]
Linter rules are all arbitrary conventions, what's the difference, please?

**Comment by ljharb** [comment]
Linter rules are arbitrary conventions you don’t have to remember.

Your concept of a debug symbol is a convention that can’t be linted for.

Regarding your concern about conflict, you can use any string as a symbol description - do you think it’s likely that a library would have a symbol with the same description as `Symbol('Lcfvs’ fancy debug symbol')`, for example? You could even (probably should) define it once in a central place and export/import it, which would have the side benefit of removing this rule’s warning at all the usage sites anyways.

**Comment by Lcfvs** [comment]
The idea is to describe it in the object property value (labelled by the symbol), then it's explicit too.

I agree, `Symbol('Lcfvs’ fancy debug symbol')`, doens't have a lot of conflict risks... but not 0.

An example? a messages dictionary, where the messages are named by their content symbols.

**Comment by Lcfvs** [comment]
My proposal is general idea.

My examples are about a "module identifier", but maybe, instead an error, the rule can only warn when it matches that pairs style, by a `Empty symbols should only be used to metadata descriptions`

**Comment by ljharb** [comment]
Right - but you're claiming that it's good to have an empty symbol as the key for that purpose; I don't think that's universal, and I think that using an empty symbol for that is a *bad practice* - it should have a clear description.

**Comment by platinumazure** [comment]
I think at the very least, this shouldn't be a new default behavior and would need to be an option.

Personally, the proposal feels very specific and as @ljharb notes, is not a universal convention by any means. So I think making any modification to this rule in ESLint core would just confuse other users.

You could use [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer) to create a custom rule based off of this rule, and use the `filterReports` method to remove lint messages that correspond to property keys which have a Symbol with description as the property value.

**Comment by Lcfvs** [comment]
Ok, ok, thanks all.
